### DISCUSSION TITLE
Cache invalidation for future dated (pending) elements

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#9170] [Ideas] [answered]
> Created by: Timothy Kelty (timkelty@gmail.com) on 2015/04/02 12:33:05 +0000
> Votes at time of UserVoice import: 6

-----------------------------------

Wherever I'm using the cache tag, I find myself eventually having to use this solution, if a client tries to use future-dated entries: http://craftcms.stackexchange.com/questions/2610/how-can-i-cache-a-list-of-entries-with-pending-postdates-and-have-it-automatical

Seems like there should be a way to bake this into Craft's native cache invalidation routines so I didn't have to do the same thing on every template.

**Comment by MoritzLost** [comment]
I just came across this issue in a project, +1 for the suggestion to handle this natively. Would be great if Craft could handle this scenario, even though I'm not sure how this could be solved.

Another related issue: What about creating new entries that would match an element query inside a cache tag? For example, let's say I have a calendar page that lists entries chronologically and I'm caching the entire entry list. Saving any entry in that list will invalidate the cache. But adding a new one that should be displayed in that list will not.

Again, not sure if there is a good solution to this issue. Maybe keep track of the element query somehow and execute it after every save to see if the results changed? Though maybe that's too complicated.

How are y'all handling those issues right now? Especially in scenarios with a lot of nested component templates, where the `{% cache %}` tag isn't necessarily aware of the element queries that are going to be executed inside of it (like the linked answer assumes)?

**Reply by brandonkelly** [reply]
We are moving support for future entry publishing/expiring to scheduled jobs in Craft 5, which will resolve this (see #8847). In the meantime, [this module](https://github.com/putyourlightson/craft-element-status-events) can be used to accomplish it.

**Reply by MoritzLost** [reply]
That's great, thanks @brandonkelly! Looking forward to this in Craft 5. In the meantime I'll have a look at that plugin.

**Reply by mikesnoeren** [reply]
> We are moving support for future entry publishing/expiring to scheduled jobs in Craft 5

I couldn’t find any documentation on this. Has this feature been released, or is it still planned for release?

**Reply by MoritzLost** [reply]
@brandonkelly I was also looking for a solution on this – is something like this already implemented, or planned?

**Comment by FreekVR** [comment]
This problem affects headless sites heavily.

Because the GraphQL cache isn't invalidated when a scheduled post should appear, the post in question will never show up until something else in the CMS changes.

**Comment by stephenmatthews** [comment]
Hey @brandonkelly - I was just wondering if there was an update on this?

**Comment by brandonkelly** [comment, marked as answer]
Just [PR’d a solution for this](https://github.com/craftcms/cms/pull/17024) for Craft 5.7: there’s a new `staticStatuses` config setting, which can be used to opt into having entry statuses stored statically and only updated on entry save; plus an `update-statuses` CLI command that can be hit with a Cron job, which updates any entries with out-of-date statuses, triggering `save` events and clearing caches in the process. (Plus a new `Entry::$oldStatus` property that can be checked from `save` event handlers to see if the status just changed.)

**Reply by brandonkelly** [reply]
Craft 5.7.0 is out with that change.

**Reply by mikesnoeren** [reply]
Great news! Do you think it could work without a cron job? Maybe by queuing a job with a delay (like `currentTime - postDate +1s`)? Not sure if that's feasible, but thought I'd throw it out there.

**Reply by brandonkelly** [reply]
We’re going to look into whether we can schedule the command using EventBridge for Craft Cloud.

Not all queue services allow delayed jobs, and those that do usually have a maximum delay. SQS for example, only allows a delay of up to 15 minutes.
